resolver/resolver.go

```diff
@@ -192,6 +193,14 @@ func (r *Resolver) servicesWithRule(ctx *context, files []*File) []*Service {
 }
 
 func (r *Resolver) validateServiceDependency(ctx *context, service *Service) {
+	// Avoid an infinite loop caused by the cyclic dependency
```

Comment:
We must stop here since if we proceed; it may go into an infinite loop if the graph contains cyclic dependency. This is why I introduced `CyclicDependencyError` in this PR.